File: api/v1/route.go
// (c) Siemens AG 2023
//
// SPDX-License-Identifier: MIT

package v1

import (
	"encoding/json"
	"fmt"
	"net"

	"github.com/siemens/ghostwire/v2/network"
)

// ipvxRoutes is the set of IPv4 and IPv6 routes discovered in a single network
// namespace.
type ipvxRoutes struct {
	IPv4 routes `json:"ipv4"`
	IPv6 routes `json:"ipv6"`
}

// routes represents a JSON marshallable list of routes (for a single address
// family)
type routes []network.Route

// route describes a single route and is marshallable to JSON.
type route struct {
	Type           string                `json:"type"`
	Family         network.AddressFamily `json:"family"`
	Destination    net.IP                `json:"destination"`
	DestinationLen int                   `json:"destination-prefixlen"`
	Index          int                   `json:"index,omitempty"`
	NifRef         string                `json:"network-interface-idref,omitempty"`
	NextHop        net.IP                `json:"next-hop,omitempty"`
	Preference     string                `json:"preference"`
	Priority       int                   `json:"priority"`
	Table          int                   `json:"table"`
}

// MarshalJSON marshals a list of routes into JSON format.
func (r routes) MarshalJSON() ([]byte, error) {
	rts := make([]route, 0, len(r))
	for _, rt := range r {
		rts = append(rts, route{
			Type:           rt.Type.String(),
			Family:         rt.Family,
			Destination:    rt.Destination.IP,
			DestinationLen: rt.DestinationPrefixLen,
			Index:          rt.Index,
			NifRef:         nifID(rt.Nif),
			NextHop:        rt.NextHop,
			Preference:     fmt.Sprintf("%02b", rt.Preference&0x03),
			Priority:       rt.Priority,
			Table:          rt.Table,
		})
	}
	return json.Marshal(rts)
}

File: webui/src/models/gw/route.ts
// (c) Siemens AG 2023
//
// SPDX-License-Identifier: MIT

import { AddressFamily } from './address'
import { NetworkInterface } from './nif';

export enum RouteType {
    Unicast = "unicast",
    Local = "local",
    Broadcast = "broadcast",
    Multicast = "multicast",
}

export interface IpRoute {
    destination: string
    prefixlen: number
    family: AddressFamily
    index: number
    nif: NetworkInterface
    nexthop?: string
    preference: string
    priority: number
    table: number
    type: RouteType
}

export const RouteTableLocal = 255

const routeString = (rt: IpRoute) =>
    `${rt.destination}/${rt.prefixlen.toString().padStart(3, '0')}+${rt.priority.toString().padStart(3, '0')}`

export const orderRoutes = (rt1: IpRoute, rt2: IpRoute) => {
    if (rt1.family !== rt2.family) {
        return rt1.family - rt2.family
    }
    return routeString(rt1).localeCompare(routeString(rt2))
}

export const routeKey = (rt: IpRoute) =>
    `${routeString(rt)}+${rt.nif && rt.nif.name}`
    